File: Wire-iOS/Sources/UserInterface/UserProfile/Views/ParticipantDeviceCell.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "ParticipantDeviceCell.h"
#import "ParticipantDeviceCell+Internal.h"

#import "WireSyncEngine+iOS.h"
#import "UIImage+ZetaIconsNeue.h"
#import "WireExtensionComponents.h"
@import PureLayout;
#import "NSString+Fingerprint.h"
#import "Wire-Swift.h"



@interface ParticipantDeviceCell ()

@property (strong, nonatomic) UIImageView *trustLevelImageView;

@end


@implementation ParticipantDeviceCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    
    if (self) {
        self.backgroundColor = UIColor.clearColor;
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        self.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        [self createViews];
        [self setupConstraints];
        [self setupStyle];
    }
    
    return self;
}

- (void)createViews
{
    self.nameLabel = [[UILabel alloc] initForAutoLayout];
    [self.contentView addSubview:self.nameLabel];
    
    self.identifierLabel = [[UILabel alloc] initForAutoLayout];
    [self.contentView addSubview:self.identifierLabel];
    
    self.trustLevelImageView = [[UIImageView alloc] initForAutoLayout];
    self.trustLevelImageView.contentMode = UIViewContentModeScaleAspectFit;
    self.trustLevelImageView.clipsToBounds = YES;
    [self.contentView addSubview:self.trustLevelImageView];
}

- (void)setupConstraints
{
    [self.trustLevelImageView autoSetDimensionsToSize:CGSizeMake(16, 16)];
    [self.trustLevelImageView autoPinEdgeToSuperviewEdge:ALEdgeLeading withInset:24];
    [self.trustLevelImageView autoAlignAxis:ALAxisHorizontal toSameAxisOfView:self.nameLabel];
    
    [self.nameLabel autoPinEdgeToSuperviewEdge:ALEdgeTop withInset:16];
    [self.nameLabel autoPinEdge:ALEdgeLeading toEdge:ALEdgeTrailing ofView:self.trustLevelImageView withOffset:16];
    
    [self.identifierLabel autoPinEdge:ALEdgeLeading toEdge:ALEdgeLeading ofView:self.nameLabel];
    [self.identifierLabel autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.nameLabel withOffset:0 relation:NSLayoutRelationGreaterThanOrEqual];
    [self.identifierLabel autoPinEdgeToSuperviewEdge:ALEdgeBottom withInset:16];
}

- (void)setHighlighted:(BOOL)highlighted animated:(BOOL)animated
{
    [super setHighlighted:highlighted animated:animated];
    self.backgroundColor = highlighted ? [UIColor colorWithWhite:0 alpha:0.08] : UIColor.clearColor;
}

#pragma mark - Configuration

- (void)configureForClient:(UserClient *)client
{
    NSDictionary *attributes = @{ NSFontAttributeName: self.fingerprintFont.monospaced };
    NSDictionary *boldAttributes = @{ NSFontAttributeName: self.boldFingerprintFont.monospaced };
    self.identifierLabel.attributedText = [client attributedRemoteIdentifier:attributes boldAttributes:boldAttributes uppercase:YES];
    self.nameLabel.text = client.deviceClass.uppercaseString ?: client.type.uppercaseString;
    self.trustLevelImageView.image = client.verified ? WireStyleKit.imageOfShieldverified : WireStyleKit.imageOfShieldnotverified;
}

@end

File: Wire-iOS/Sources/UserInterface/UserProfile/Views/ParticipantDeviceCell.swift
//
// Wire
// Copyright (C) 2019 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

final class ParticipantDeviceCell: UITableViewCell {

    private var nameLabel: UILabel!
    private let boldFingerprintFont: UIFont = .smallSemiboldFont
    private let fingerprintFont: UIFont = .smallFont
    private var identifierLabel: UILabel!
    private var trustLevelImageView: UIImageView!

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)

        backgroundColor = UIColor.clear
        selectionStyle = .none
        accessoryType = .disclosureIndicator

        createViews()
        setupConstraints()
        setupStyle()
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func createViews() {
        nameLabel = UILabel()
        contentView.addSubview(nameLabel)

        identifierLabel = UILabel()
        contentView.addSubview(identifierLabel)

        trustLevelImageView = UIImageView()
        trustLevelImageView.contentMode = .scaleAspectFit
        trustLevelImageView.clipsToBounds = true
        contentView.addSubview(trustLevelImageView)
    }

    private func setupConstraints() {
        [trustLevelImageView,
         nameLabel,
         identifierLabel].forEach() { $0.translatesAutoresizingMaskIntoConstraints = false }

        var constraints: [NSLayoutConstraint] = []

        constraints += trustLevelImageView.setDimensions(length: 16, activate: false)
        constraints.append(trustLevelImageView.pinToSuperview(anchor: .leading, inset: 24, activate: false))
        constraints.append(trustLevelImageView.pin(to: nameLabel, axisAnchor: .centerY, activate: false))

        constraints.append(nameLabel.pinToSuperview(anchor: .top, inset: 16, activate: false))
        constraints.append(nameLabel.leadingAnchor.constraint(equalTo: trustLevelImageView.trailingAnchor, constant: 16))

        constraints.append(identifierLabel.pin(to: nameLabel, anchor: .leading, activate: false))
        constraints.append(identifierLabel.topAnchor.constraint(greaterThanOrEqualTo: nameLabel.bottomAnchor))
        constraints.append(identifierLabel.pinToSuperview(anchor: .bottom, inset: 16, activate: false))

        NSLayoutConstraint.activate(constraints)
    }

    override func setHighlighted(_ highlighted: Bool, animated: Bool) {
        super.setHighlighted(highlighted, animated: animated)
        backgroundColor = highlighted ? UIColor(white: 0, alpha: 0.08) : UIColor.clear
    }

    // MARK: - Configuration

    @objc(configureForClient:)
    func configure(for client: UserClientType & UserClientTypeAttributedString) {
        let attributes: [NSAttributedString.Key: AnyObject] = [NSAttributedString.Key.font: fingerprintFont.monospaced()]
        let boldAttributes: [NSAttributedString.Key: AnyObject] = [NSAttributedString.Key.font: boldFingerprintFont.monospaced()]

        identifierLabel.attributedText = client.attributedRemoteIdentifier(attributes, boldAttributes: boldAttributes, uppercase: true)
        nameLabel.text = client.deviceClass?.uppercased() ?? client.type.uppercased()


        trustLevelImageView.image = client.verified ? WireStyleKit.imageOfShieldverified : WireStyleKit.imageOfShieldnotverified
    }

    private func setupStyle() {
        nameLabel.textColor = .from(scheme: .textForeground)
        nameLabel.font = .smallSemiboldFont

        identifierLabel.textColor = .from(scheme: .textForeground)
    }
}
